File: mobile/modules/instax-printer/android/src/main/java/expo/modules/instaxprinter/InstaxPrinterModule.kt
package expo.modules.instaxprinter

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException

// Stub implementation. Replace the bodies with calls into Fujifilm's Instax SDK
// (or your own BLE code) once the SDK aar is dropped into android/libs.

class InstaxPrinterModule : Module() {
    private var state: String = "disconnected"
    private var deviceName: String? = null
    private var filmRemaining: Int = 8

    private fun statusMap(): Map<String, Any?> = mapOf(
        "state" to state,
        "deviceName" to deviceName,
        "batteryLevel" to 0.82,
        "filmRemaining" to filmRemaining,
    )

    override fun definition() = ModuleDefinition {
        Name("InstaxPrinter")

        AsyncFunction("scanForDevices") { _: Int ->
            // TODO: Use BluetoothLeScanner with the Instax service UUID filter.
            listOf(
                mapOf("id" to "instax-mini-link-2", "name" to "INSTAX-MINI Link 2", "rssi" to -52),
                mapOf("id" to "instax-square-link", "name" to "INSTAX SQUARE Link", "rssi" to -67),
            )
        }

        AsyncFunction("connect") { deviceId: String? ->
            // TODO: BluetoothGatt connectGatt + Instax pairing handshake.
            state = "connected"
            deviceName = deviceId ?: "INSTAX (stub)"
            statusMap()
        }

        AsyncFunction("disconnect") {
            state = "disconnected"
            deviceName = null
            statusMap()
        }

        AsyncFunction("getStatus") {
            statusMap()
        }

        AsyncFunction("print") { imageUri: String ->
            // TODO: Decode imageUri to a Bitmap, scale to the target Instax format
            // (mini = 600x800 px), and call the SDK send-print method. Suspend on
            // the printer's print-finished callback before returning.
            if (state != "connected") {
                throw CodedException("NotConnected", "Printer not connected", null)
            }
            state = "printing"
            Thread.sleep(1500)
            filmRemaining = (filmRemaining - 1).coerceAtLeast(0)
            state = "connected"
            mapOf("jobId" to "stub-${System.currentTimeMillis()}")
        }
    }
}

File: mobile/modules/instax-printer/ios/InstaxPrinterModule.swift
import ExpoModulesCore

// Stub implementation. Replace the bodies of `connect`, `print`, etc. with calls
// into Fujifilm's Instax SDK (or your own CoreBluetooth code) once the SDK is
// linked into the Xcode project.

public class InstaxPrinterModule: Module {
    private var state: String = "disconnected"
    private var deviceName: String? = nil
    private var filmRemaining: Int = 8

    private func statusDict() -> [String: Any] {
        var dict: [String: Any] = ["state": state]
        if let n = deviceName { dict["deviceName"] = n }
        dict["batteryLevel"] = 0.82
        dict["filmRemaining"] = filmRemaining
        return dict
    }

    public func definition() -> ModuleDefinition {
        Name("InstaxPrinter")

        AsyncFunction("scanForDevices") { (_ timeoutMs: Int) -> [[String: Any]] in
            // TODO: Replace with CoreBluetooth scan filtered to Fujifilm Instax service UUIDs.
            return [
                ["id": "instax-mini-link-2", "name": "INSTAX-MINI Link 2", "rssi": -52],
                ["id": "instax-square-link", "name": "INSTAX SQUARE Link", "rssi": -67],
            ]
        }

        AsyncFunction("connect") { (deviceId: String?) -> [String: Any] in
            // TODO: Resolve a CBPeripheral by id and connect via the Fujifilm SDK.
            self.state = "connected"
            self.deviceName = deviceId ?? "INSTAX (stub)"
            return self.statusDict()
        }

        AsyncFunction("disconnect") { () -> [String: Any] in
            self.state = "disconnected"
            self.deviceName = nil
            return self.statusDict()
        }

        AsyncFunction("getStatus") { () -> [String: Any] in
            return self.statusDict()
        }

        AsyncFunction("print") { (imageUri: String) -> [String: String] in
            // TODO: Decode imageUri (file:// or https://) into a UIImage, downscale to
            // the target Instax format (mini = 600x800 px @ 318 dpi), then call the SDK
            // print method. Wait for completion or progress events before resolving.
            guard self.state == "connected" else {
                throw Exception(name: "NotConnected", description: "Printer not connected")
            }
            self.state = "printing"
            // simulate latency
            Thread.sleep(forTimeInterval: 1.5)
            self.filmRemaining = max(0, self.filmRemaining - 1)
            self.state = "connected"
            return ["jobId": "stub-\(Int(Date().timeIntervalSince1970))"]
        }
    }
}
